xdp/src/netlink.rs

```diff
@@ -0,0 +1,601 @@
+#![allow(clippy::arithmetic_side_effects)]
+
+use {
+    libc::{
+        getsockname, nlattr, nlmsgerr, nlmsghdr, recv, send, setsockopt, sockaddr_nl, socket,
+        AF_INET, AF_INET6, AF_NETLINK, NDA_DST, NDA_LLADDR, NETLINK_EXT_ACK, NETLINK_ROUTE,
+        NLA_ALIGNTO, NLA_TYPE_MASK, NLMSG_DONE, NLMSG_ERROR, NLM_F_DUMP, NLM_F_MULTI,
+        NLM_F_REQUEST, NUD_PERMANENT, NUD_REACHABLE, NUD_STALE, RTA_DST, RTA_GATEWAY, RTA_IIF,
+        RTA_OIF, RTA_PREFSRC, RTA_PRIORITY, RTA_TABLE, RTM_GETNEIGH, RTM_GETROUTE, RTM_NEWNEIGH,
+        RTM_NEWROUTE, RT_TABLE_MAIN, SOCK_RAW, SOL_NETLINK,
+    },
+    std::{
+        collections::HashMap,
+        io, mem,
+        net::{IpAddr, Ipv4Addr, Ipv6Addr},
+        os::fd::{AsRawFd, FromRawFd, OwnedFd},
+        ptr, slice,
+    },
+    thiserror::Error,
+};
+
+const NLA_HDR_LEN: usize = align_to(mem::size_of::<nlattr>(), NLA_ALIGNTO as usize);
+
+pub struct NetlinkSocket {
+    sock: OwnedFd,
+    _nl_pid: u32,
+}
+
+impl NetlinkSocket {
+    fn open() -> Result<Self, io::Error> {
+        // Safety: libc wrapper
+        let sock = unsafe { socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE) };
+        if sock < 0 {
+            return Err(io::Error::last_os_error());
+        }
+        // SAFETY: `socket` returns a file descriptor.
+        let sock = unsafe { OwnedFd::from_raw_fd(sock) };
+
+        let enable = 1i32;
+        // Safety: libc wrapper
+        unsafe {
+            setsockopt(
+                sock.as_raw_fd(),
+                SOL_NETLINK,
+                NETLINK_EXT_ACK,
+                &enable as *const _ as *const _,
+                mem::size_of::<i32>() as u32,
+            )
```

Comment:
Check for error?

Reply:
I'd rather not. I'd have to `use` c_void and socklen_t. We can elide as we elide types everywhere right? You can configure your editor to show you the types

Reply:
implemented error check